### DISCUSSION TITLE
Connecting BLE: android app <--> ESP32

### DISCUSSION THREAD

**Discussion by rodo54** [micropython/micropython#9507] [Hardware & Peripherals] [answered]
Hi guys,
So far I have programmed my robots using MicroPython on Raspberry Pi boards.
Now I try to migrate to **ESP32** (cheaper and especially faster booting). There is no mystery regarding the handling of servos and motors, but I am somewhat confused about the bluetooth link with the cell phone.
With the RP I used the **_Bluedot_** application with its corresponding module but I don't know how to solve it in the ESP32.
(_ESP32 WROOM DevKit_ with _esp32-20220618-v1.19.1_ firmware)

I have mentioned the Bluedot app, but it could be any other app suitable for eg driving a remote control car.

Could someone suggest me examples or tutorials to learn about this?
Thanks in advance.

**Comment by jimmo** [comment, marked as answer]
Bluedot (from what I can tell from the docs) communicates over rfcomm, which is Bluetooth Classic. rfcomm is essentially a serial port over Bluetooth.

MicroPython only supports Bluetooth Low Energy.

Adafruit have an Android/iOS app called "Bluefruit Connect" which includes a [controller mode](https://learn.adafruit.com/bluefruit-le-connect/controller#control-pad-2923571).

It's been a while, but I'm fairly sure that Bluefruit works over the Nordic UART Service (essentially rfcomm but for BLE). I think you should be able to use https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_peripheral.py to receive events from Bluefruit.

**Reply by rodo54** [reply]
> I'm fairly sure that Bluefruit works over the Nordic UART Service (essentially rfcomm but for BLE). I think you should be able to use https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_peripheral.py to receive events from Bluefruit. 

it really works! 
It would be very intuitive to drive the car with the accelerometer data.
I just need to decode the received binary data.
I will work on it.

**Comment by sonnny** [comment]
Some links of my experimentation on Android/BLE and also android/WiFi

[ble esp32 notification/android](https://crus.in/codes/ble_notification.txt)

[youtube demo ble notification esp32, source in the description](https://www.youtube.com/watch?v=NCaK9VPx_S8)

[youtube demo ble esp32 android, source in the description](https://www.youtube.com/shorts/nx_TeehNUPc)

[youtube demo pico w/android source in the description](https://www.youtube.com/watch?v=Zm6jv7Z-n2M)

**Comment by rodo54** [comment]
I have just completed the documentation and it is available at https://github.com/micropython/micropython/discussions/9925
:)
